### DISCUSSION TITLE
How to use CompositeDevice and SPI

### DISCUSSION THREAD

**Discussion by karwenzman** [gpiozero/gpiozero#938] [Q&A] [answered]
Hello!
Is this a place where I can ask for some support?

I have read a lot of articles and did some research on 'gpiozero', but I need help using the class 'CompositeDevice' and SPI.

- Could you share an example, how to setup SPI-devices with gpiozero? I was able to use SPIDEV, but I did not figure out how to use it with gpiozero. 
- Here is the link to this thread: i would like to create a class using CompositeDevice as base class. I was playing around with some classes like BUTTON, LED; OutputDevice. That did work fine. However, i was not able to add a SPI device.

[SPI](https://gpiozero.readthedocs.io/en/stable/api_spi.html#spidevice):
>"_A hardware based implementation. This is only available when the SPI kernel module is loaded, and the Python spidev library is available. It can only use specific pins for SPI communication (GPIO11=clock, GPIO10=MOSI, GPIO9=MISO, while GPIO8 is select for device 0 and GPIO7 is select for device 1). However, it is extremely fast and works with all devices._"

My intention is to write code for a little board MFRC522. RFID-Reader with SPI. 

Some help here will be much appreciated.

```python
class BreadboardRC522(gpiozero.CompositeDevice):
    def __init__(self, *args, _order=None, pin_factory=None, **kwargs):
        # loading '__init__' of base class 'CompositeDevice'
        super().__init__(*args, _order=_order, pin_factory=pin_factory, **kwargs)
```

```python
def main():
    rc522 = BreadboardRC522(
        rst = gpiozero.OutputDevice(pin=16, initial_value=False),
        irq = gpiozero.Button(pin=17, pull_up=True, bounce_time=0.02),
        led = gpiozero.LED(pin=20, initial_value=False),
        )
```

_Originally posted by @karwenzman in https://github.com/gpiozero/gpiozero/issues/522#issuecomment-640835407_

**Comment by waveform80** [comment, marked as answer]
The docs are admittedly not great on this front and at the moment we've only got (effectively) one example of an SPI driven device in the API which is the MCP3xxx series of ADCs. You can find the code for these in [gpiozero/spi_devices.py](https://github.com/gpiozero/gpiozero/blob/master/gpiozero/spi_devices.py), but the functional bits for the ADCs are spread across several classes with a lot of abstraction there (because there's so much similarity between all those chips), and furthermore those classes are quite complex because we pretend that each ADC channel is a separate device (it's debatable whether this was a wise choice or not - I'm still not sure).

So, here's a stab at a walkthrough for making SPI interfaces in GPIO Zero -- maybe this'll be a template for a chapter in the docs at some point.

Here's the stuff you need know:

* Anything derived from `SPIDevice` has an internal `self._spi` object which represents the SPI interface it's talking to. This implements the [SPI abstract class](https://gpiozero.readthedocs.io/en/stable/api_pins.html#gpiozero.SPI) with various methods like `read`, `write`, `transfer` and properties like `clock_mode` and so on -- all the stuff you'd expect in an SPI interface, basically.

* You'll use the `self._spi` object to talk to your SPI device. Don't worry about how it's constructed or what's actually underneath the covers -- it'll be on the pins (or port/device) requested when you constructed the class, and it'll be the best interface GPIO Zero can find (e.g. if it's on the kernel-supported SPI pins and the `spidev` library is importable we'll use that, otherwise we'll emulate an interface with bit-banging).

* The various methods like `transfer` expect (and return) a list of integers. This is because, theoretically SPI interfaces can use an arbitrary number of bits per word. The *vast* majority of the time they use 8-bits per word so most interfaces just use ASCII bytes, but there's nothing stopping devices from demanding 10-bits per word (for example). As the GPIO Zero interface is generic, we simply deal in sequences of integers, but do note that byte-strings in Python 3 iterate as *sequences of integers* so actually you can just pass a byte-string to methods like `write` or `transfer`, or just do `bytes(seq)` on the result if you want to handle byte-strings.

* You also have a couple of utility methods in `SPIDevice` which are `_int_to_words` and `_words_to_int` that ease conversion between multi-byte integer values (of arbitrary size) and the sequences of integers up to the value of the `bits_per_word` attribute on the internal `self._spi` interface. These are particularly useful on things like ADCs but are probably quite useful in other applications too.

Now, here's the stuff you need to do:

* Implement something derived from `SPIDevice` which provides a `value` property that, when read, returns the "value" of the device (and, if it's a "writeable" device -- something like an output -- changes the "value" of the device when written to).

* Erm ... that's about it. Add a `__repr__` method if you want a friendly command line? Maybe some tests? Some documentation? All that's stuff that we'd want for inclusion of your SPI class in GPIO Zero, but if it's just for you, then it's up to you.

Here's how I'd go about dealing with the [MCP3008](https://learn.adafruit.com/raspberry-pi-analog-to-digital-converters/mcp3008) if I were handling it as a standalone device in GPIO Zero with no other similar devices to deal with. It's an 8-channel ADC so let's make a class which has a "channels" attribute which determines which channels we want to return. The "value" of the device will be a tuple of each channel selected (personally I think this is probably how I *should've* handled the MCP3xxx series instead of a device-per-channel but that does complicate the source/values interface ... maybe something we'll re-visit in 2.x).

1. Let's start with the constructor. Pretty trivial stuff here: we'll take a default for our "channels" attribute, and everything else is SPI pins arguments that we'll chuck at the super-class to handle.

```python3
from gpiozero import SPIDevice

class MCP3008ADC(SPIDevice):
    def __init__(self, channels=None, **spi_args):
        if channels is None:
              channels = list(range(8))
        self.channels = channels
        super().__init__(**spi_args)
```

2. Now we need to deal with the "channels" attribute - just some basic type-checking/conversion here:

```python3
    @property
    def channels(self):
        return self._channels

    @channels.setter
    def channels(self, value):
          channels = [int(i) for i in value]
          if any(not 0 <= c < 8 for c in channels):
              # Should probably mention *which* channel here, but I'm lazy
              raise ValueError('invalid channel!')
          self._channels = channels
```

3. Now the meat! Using the `self._spi` interface to ask the device what the values of various channels are. This is the point where you want to go and read the [datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21295d.pdf) for your device to figure out what its protocol is. In our case, the MCP3008 is a 10-bit device and its protocol is:
  - write a "1" start-bit, followed by a differential bit (always 0 here), and 3-bits of channel
  - then on the reading side there's an unused bit (basically a pause while it reads the channel), a "0" start-bit, followed by 10-bits of the value.

4. Some SPI implementations only support byte-sized words, so we'll assume we're working with that (it's the default anyway). The chip-select must be active throughout so we need to do this in a single transfer. Hence we want to call `self._spi.transfer` with something like `[0b00000001, 0b0CCCx0RR, 0bRRRRRRRR]` where the Cs are the channel bits, and the Rs are bits we want to read (but we don't care what we transmit) and the x is a bit that's literally unused:

```python3
    @property
    def value(self):
        result = []
        for c in self._channels:
            reading = self._spi.transfer([1, c << 4, 0])
            reading = reading[-2:] # only want the last two bytes
            reading = self._words_to_int(reading, 10) # convert them to a 10-bit int
            # In keeping with convention we'll convert the reading to a 0..1 value
            reading = reading / (2 ** 10)
            result.append(reading)
        return tuple(result)
```

And that's it. You can construct an instance of this class like so:

```python3
adc = MCP3008ADC(channels=[0, 1, 2], port=0, device=0)
```

Or you can specify SPI pins explicitly:

```python3
adc = MCP3008ADC(channels=[0, 1, 2], clock_pin=11,
    mosi_pin=10, miso_pin=9, select_pin=8)
```

Then you can read the `value` property to query the device. Or you could incorporate that device into a `CompositeDevice`, etc. etc.

**Reply by karwenzman** [reply]
Thanks a lot for this fine comment!
